src/pytorch_lightning/core/mixins/device_dtype_mixin.py

```diff
@@ -57,7 +50,7 @@ def device(self) -> Union[str, torch.device]:
 
         return device
 
-    def to(self, *args: Any, **kwargs: Any) -> Self:
+    def to(self, *args: Any, **kwargs: Any) -> TDeviceDtypeModuleMixin:
```

Comment:
The first set of mypy errors goes away if you also annotate `self` argument as `TDeviceDtypeModuleMixin`. However, this is tricky, since as @awaelchli said (https://github.com/Lightning-AI/lightning/pull/13541#discussion_r914449104), annotating `self` is almost always a bad idea. So, although I'm not a fan, I'd vote for annotating these with some sort of `type: ignore`s. If you do, then the errors manifested in the `trainer/connectors/logger_connector/result.py` will probably go away as well.

Does anyone have any other opinions? @awaelchli @carmocca

Reply:
The issue with `typing.Self` is that it is only in 3.11 and while it is available in `typing_extensions`, mypy does not work with that (https://github.com/python/mypy/issues/11871). And since we're using mypy as a checker of choice, we're stuck with not being able to use `typing_extensions.Self`

Reply:
This is tricky. My recommendation is that we adhere to what's correct in the ideal typing world and ignore the warnings provided by mypy, as they are caused by current limitation of the software.

Once they add support for it, mypy should start complaining that the ignores are no longer required, so we will only need to remove the ignore then, and not modify the source code.

Reply:
@carmocca and @otaj I have made this change. However, the issue with `logger_connector/result.py` is seen with this change

```
src/pytorch_lightning/trainer/connectors/logger_connector/result.py:202: error: Definition of "float" in base class "Metric" is incompatible with definition in base class "DeviceDtypeModuleMixin"  [misc]
```

Reply:
Well, as @awaelchli and @carmocca said, let's stick to what is correct in the ideal typing world. That means annotating with `Self` from `typing_extensions`, although I know I recommended a different route at first, @carmocca's comment absolutely makes sense to me - let's do what's right and wait for mypy to catch up